eventlet/green/nanomsg.py

```diff
@@ -0,0 +1,56 @@
+# coding: utf-8
+"""The :mod:`nanomsg` module wraps the :class:`Socket`
+found in :mod:`nanomsg-python <nanomsg>` to be non blocking.
+"""
+
+__nanomsg__ = __import__('nanomsg')
+from eventlet.patcher import slurp_properties
+from eventlet.hubs import trampoline
+from nanomsg import wrapper
+
+__nanomsg__.__all__ = ['wrapper', 'NanoMsgError', 'NanoMsgAPIError', 'Device', 'Socket']
+
+for name, value in wrapper.nn_symbols():
+    if name.startswith('NN_'):
+        name = name[3:]
+    __nanomsg__.__all__.append(name)
+
+__patched__ = ['Socket']
+slurp_properties(__nanomsg__, globals(), ignore=__patched__)
+
+
+class Socket(__nanomsg__.Socket):
+
+    def recv(self, buf=None, flags=0):
+        """Recieve a message."""
+
+        # Wait for the receive file descriptor using eventlet
+        while True:
+
+            # Call the eventlent hub trampoline function to wait for a notification on the
+            # receive file descriptor.
+            trampoline(self.recv_fd, read=True)
+
+            # Don't allow nn_recv() to block
+            flags = flags | __nanomsg__.DONTWAIT
+
+            if buf is None:
+                rtn, out_buf = wrapper.nn_recv(self.fd, flags)
+            else:
+                rtn, out_buf = wrapper.nn_recv(self.fd, buf, flags)
```

Comment:
I'm sorry, I'm unfamiliar with nanomsg and its implementation, but it's atypical for a Python function to be "overloaded" in such a way as to change the meaning of a positional argument.

Is it really true that if you pass `nanomsg.wrapper.nn_recv()` two positional arguments, the second one is `flags`, but if you pass three, the second one is `buf`?

Reply:
I agree this is very odd, but python-nanomsg does this internally:

```python
def nn_recv(socket, *args):
    "receive a message"
    if len(args) == 1:
        flags, = args
        pointer = ctypes.c_void_p()
        rtn = _nn_recv(socket, ctypes.byref(pointer), ctypes.c_size_t(-1),
                       flags)
        if rtn < 0:
            return rtn, None
        else:
            return rtn, _create_message(pointer.value, rtn)
    elif len(args) == 2:
        msg_buf, flags = args
        mv_buf = memoryview(msg_buf)
        if mv_buf.readonly:
            raise TypeError('Writable buffer is required')
        rtn = _nn_recv(socket, ctypes.addressof(msg_buf), len(mv_buf), flags)
        return rtn, msg_buf
```

Reply:
Wow! A reader of your `Socket.recv()` method would certainly benefit from an explanatory comment. You don't have to paste their code, just explain that they really do take different branches depending on the number of passed arguments.